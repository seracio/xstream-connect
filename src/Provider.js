// @flow
import PropTypes from 'prop-types';
import React from 'react';

class Provider extends React.Component {

  constructor(props, context) {
    super(props, context);
    this.store = props.store;
  }

  getChildContext() {
    return { store: this.store };
  }

  render(){
    return React.Children.only(this.props.children);
  }
}

Provider.propTypes = {
  store: PropTypes.object.isRequired,
};

Provider.childContextTypes = {
  store: PropTypes.object.isRequired,
};

export default Provider;
