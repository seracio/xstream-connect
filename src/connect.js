// @flow
import _ from 'lodash/fp';
import React from 'react';
import xs from 'xstream';

const mapWithIndex = _.map.convert({cap: false});

const connect = storeToPropsFunc => WrappedComponent => {

  if(typeof(storeToPropsFunc) !== 'function'){
    throw new Error('xstream-connect: connect needs a function storeToPropsFunc as parameter');
  }

  class Connect extends React.Component {
    constructor(props, context) {
      super(props, context);
      // flag
      this.go = false;
      // the fragment of the store we'll listen
      this.fragment = storeToPropsFunc(this.context.store);
      // order
      // needed for the listen method
      this.order = _.keys(this.fragment);
      // initiate the state
      // to null
      this.state = _.mapValues(
        _.constant(null),
        this.fragment
      );
    }

    componentDidMount(){
      this.listen();
    }

    listen() {
      xs.combine(..._.map(key => this.fragment[key], this.order))
        .addListener({
          next: values => {
            this.go = true;
            const state = _.flow(
              mapWithIndex((value, index) => ({ key: this.order[index], value })),
              _.keyBy(_.get('key')),
              _.mapValues(_.get('value'))
            )(values);
            this.setState(state);
          }
        });
    }

    render() {
      const propsToTransfer = {...this.props, ...this.state};
      return this.go && <WrappedComponent {...propsToTransfer}/>;
    }
  }
  Connect.contextTypes = {
    store: React.PropTypes.object.isRequired,
  };

  return Connect;
};

export default connect;
