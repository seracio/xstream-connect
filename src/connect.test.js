import React from 'react';
import renderer from 'react-test-renderer';
import xs from 'xstream';
import connect from './connect';
import Provider from './Provider';

test('', () => {

});