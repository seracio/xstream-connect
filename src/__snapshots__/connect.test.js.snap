exports[`test Connect component: fragment attribute should contains keys specified by the storeToPropsFunc parameter 1`] = `
<div>
  world
</div>
`;

exports[`test Connect component: should only render the WrappedComponent when go === true 1`] = `
<div>
  hello
</div>
`;

exports[`test Connect component: should render false when state.go === false 1`] = `null`;
